modules/ephemeris/src/main/scala/gem/horizons/HorizonsEphemerisUpdater.scala

```diff
@@ -0,0 +1,191 @@
+                   user: User[_],
+                   log:  Log[ConnectionIO],
+                   xa:   Transactor.Aux[IO, Unit]) {
+
+  import HorizonsEphemerisUpdater._
+
+  /** Constructs an action that when run will provide the current context
+    * information for an ephemeris.  This information can be used to determine
+    * whether an update is needed, to see when the last update was performed,
+    * and when the last update check happened.
+    */
+  def report(key:  EphemerisKey.Horizons, site: Site): IO[Context] =
+
+    for {
+
+      meta <- log.log(user, s"EphemerisDao.selectMeta($key, $site)") {
+                EphemerisDao.selectMeta(key, site)
+              }.transact(xa)
+
+      rnge <- log.log(user, s"EphemerisDao.selectTimes($key, $site)") {
+                EphemerisDao.selectTimes(key, site)
+              }.transact(xa)
+
+      soln <- log.log(user, s"HorizonsSolutionRefQuery($key).lookup") {
+                HorizonsSolutionRefQuery(key).lookup.liftIO[ConnectionIO]
+              }.transact(xa)
+
+    } yield Context(key, site, meta, rnge, soln)
+
+
+  /** Constructs an action that when run will insert a new ephemeris or update
+    * an existing one if necessary.
+    */
+  def update(key: EphemerisKey.Horizons, site: Site): IO[Unit] =
+
+    for {
+
+      time <- InstantMicros.now
+      sem  <- Semester.current(site)
+      ctx  <- report(key, site)
+      _    <- (if (ctx.isUpToDateFor(sem))
+                 recordUpdateCheck(ctx, time)
+               else
+                 doUpdate(ctx, time, sem)
+              ).transact(xa)
+
+    } yield ()
+
+
+  private def logMessage(msg: String): ConnectionIO[Unit] =
+    log.log(user, msg)(().pure[ConnectionIO])
+
+  private def insertMeta(key: EphemerisKey.Horizons, site: Site, m: EphemerisMeta): ConnectionIO[Unit] =
+    log.log(user, s"insertMeta($key, $site, $m)") {
+      EphemerisDao.insertMeta(key, site, m).void
+    }
+
+  private def updateMeta(key: EphemerisKey.Horizons, site: Site, m: EphemerisMeta): ConnectionIO[Unit] =
+    log.log(user, s"updateMeta($key, $site, $m)") {
+      EphemerisDao.updateMeta(key, site, m).void
+    }
+
+  private def streamEphemeris(
+                key:      EphemerisKey.Horizons,
+                site:     Site,
+                semester: Semester): Stream[ConnectionIO, Ephemeris.Element] = {
+
+    val qs = HorizonsEphemerisQuery.pagingSemester(key, site, semester, StepSize, Padding)
+
+    qs.foldMap(_.streamEphemeris)
+      .through(standardVelocityCompression)
+      .translateSync(λ[IO ~> ConnectionIO](_.liftIO[ConnectionIO]))
+  }
+
+  private def recordUpdateCheck(ctx:  Context,
+                                time: InstantMicros): ConnectionIO[Unit] =
+
+    for {
+      _ <- logMessage(s"${ctx.key}@${ctx.site} already up-to-date. Record update check.")
+      _ <- ctx.meta.fold(().pure[ConnectionIO]) { m =>
+            updateMeta(ctx.key, ctx.site, EphemerisMeta.lastUpdateCheck.set(time)(m))
+          }
```

Comment:
There should be a combinator for this. You can say `OptionT(ctx.meta).semiflatMap(m => ...).value` but it's not really much better.